### PR TITLE
Update test case to make memory hotplug/unplug when VM is paused

### PR DESCRIPTION
**Pull request by xutian**
cleanup code execute logical and made a small update to ensure VM is paused status when execute test case "hotplug_memory.during.pause_vm.hotplug.backend_file.2M_hugepage.policy_bind.two".

### CODE REVIEW

**Comment by xutian** [comment]
ID: 1285328

**Comment by suqinhuang** [comment]
Look ok to me

**Comment by xiagao** [comment]
Run three cases about pause_vm, all pass. But check the debug log, the steps are not in accordance with them in polarion.
Results:
01-smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.hotplug_memory.before.pause_vm.hotplug.backend_ram.policy_default.two Finshed 2016-09-28 04:33:06 2016-09-28 05:16:13 PASS 2587.48379
debug.log
04:34:33 INFO | Context: Stop the VM
04:34:33 DEBUG| (monitor qmpmonitor1) Sending command 'stop'
04:34:33 DEBUG| Send command: {'execute': 'stop', 'id': 'QHO8Md4N'}
04:34:33 INFO | Context: Stop the VM --> Verify the status of VM is 'paused'
04:34:33 INFO | Context: Stop the VM --> Verify the session has no response
04:35:40 INFO | Update MAC(00:22:19:30:7c:af)<->(10.66.144.69)IP pair into address_cache
04:39:38 INFO | Context: Resume the VM
04:39:38 DEBUG| (monitor qmpmonitor1) Sending command 'cont'
04:39:38 DEBUG| Send command: {'execute': 'cont', 'id': '3QptKqu1'}
04:39:38 INFO | Context: Resume the VM --> Verify the status of VM is 'running'
04:39:38 INFO | Context: Resume the VM --> Re-login the guest
04:39:38 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 600s)
04:40:04 INFO | Context: Run sub test 'pause_vm' before hotplug memory device --> Hotplug memory 'mem-plug' to VM
04:40:04 DEBUG| (monitor qmpmonitor1) Sending command 'object-add'
04:40:04 DEBUG| Send command: {'execute': 'object-add', 'arguments': {'id': 'mem-plug', 'qom-type': 'memory-backend-ram', 'props': {'policy': 'default', 'size': 1073741824}}, 'id': 'ZjbopWGQ'}
04:40:04 INFO | Context: Run sub test 'pause_vm' before hotplug memory device --> Hotplug memory 'mem-plug' to VM --> Get hotpluged memory address
04:40:04 DEBUG| (monitor qmpmonitor1) Sending command 'query-memory-devices'
04:40:04 DEBUG| Send command: {'execute': 'query-memory-devices', 'id': 'N1Q0O88c'}
04:40:04 DEBUG| (monitor qmpmonitor1) Response to 'query-memory-devices' (re-formated)
04:40:04 DEBUG| (monitor qmpmonitor1)        data:
04:40:04 DEBUG| (monitor qmpmonitor1)            slot: 0
04:40:04 DEBUG| (monitor qmpmonitor1)            node: 1
04:40:04 DEBUG| (monitor qmpmonitor1)            addr: 4294967296
04:40:04 DEBUG| (monitor qmpmonitor1)            memdev: /objects/mem-mem1
04:40:04 DEBUG| (monitor qmpmonitor1)            id: dimm-mem1
04:40:04 DEBUG| (monitor qmpmonitor1)            hotpluggable: True
04:40:04 DEBUG| (monitor qmpmonitor1)            hotplugged: False
04:40:04 DEBUG| (monitor qmpmonitor1)            size: 1073741824
04:40:04 DEBUG| (monitor qmpmonitor1)        type: dimm
04:40:04 INFO | Context: Run sub test 'pause_vm' before hotplug memory device --> Hotplug pc-dimm 'dimm-plug' to VM
04:40:04 DEBUG| (monitor qmpmonitor1) Sending command 'device_add'
04:40:04 DEBUG| Send command: {'execute': 'device_add', 'arguments': OrderedDict([('driver', 'pc-dimm'), ('id', 'dimm-plug'), ('memdev', 'mem-plug')]), 'id': 'uUMnHWa4'}
04:40:04 INFO | Context: Run sub test 'pause_vm' before hotplug memory device --> Verify memory info
-----------------------------------------------------------------------------------------
05-smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.hotplug_memory.after.pause_vm.hotplug.backend_ram.policy_default.two Finshed 2016-09-28 22:35:46 2016-09-28 22:44:31 PASS 524.866943
debug.log
22:35:52 INFO | Context: (run) --> Hotplug memory 'mem-plug' to VM
22:35:52 DEBUG| (monitor qmpmonitor1) Sending command 'object-add'
22:35:52 DEBUG| Send command: {'execute': 'object-add', 'arguments': {'id': 'mem-plug', 'qom-type': 'memory-backend-ram', 'props': {'policy': 'default', 'size': 1073741824}}, 'id': 'dLPQYls0'}
22:35:52 INFO | Context: (run) --> Hotplug memory 'mem-plug' to VM --> Get hotpluged memory address
22:35:52 DEBUG| (monitor qmpmonitor1) Sending command 'query-memory-devices'
22:35:52 DEBUG| Send command: {'execute': 'query-memory-devices', 'id': 'uM0EaeGz'}
22:35:52 DEBUG| (monitor qmpmonitor1) Response to 'query-memory-devices' (re-formated)
22:35:52 DEBUG| (monitor qmpmonitor1)        data:
22:35:52 DEBUG| (monitor qmpmonitor1)            slot: 0
22:35:52 DEBUG| (monitor qmpmonitor1)            node: 1
22:35:52 DEBUG| (monitor qmpmonitor1)            addr: 4294967296
22:35:52 DEBUG| (monitor qmpmonitor1)            memdev: /objects/mem-mem1
22:35:52 DEBUG| (monitor qmpmonitor1)            id: dimm-mem1
22:35:52 DEBUG| (monitor qmpmonitor1)            hotpluggable: True
22:35:52 DEBUG| (monitor qmpmonitor1)            hotplugged: False
22:35:52 DEBUG| (monitor qmpmonitor1)            size: 1073741824
22:35:52 DEBUG| (monitor qmpmonitor1)        type: dimm
22:35:52 INFO | Context: (run) --> Hotplug pc-dimm 'dimm-plug' to VM
22:35:52 DEBUG| (monitor qmpmonitor1) Sending command 'device_add'
22:35:52 DEBUG| Send command: {'execute': 'device_add', 'arguments': OrderedDict([('driver', 'pc-dimm'), ('id', 'dimm-plug'), ('memdev', 'mem-plug')]), 'id': '9XuZ0PeR'}
22:35:52 INFO | Context: (run) --> Verify memory info
22:35:52 WARNI| Unable to read serial console
22:35:52 WARNI| Unable to read serial console
22:35:52 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 240s)
22:36:03 DEBUG| Login timeout expired    (output: '')
22:36:43 DEBUG| Client said 'connection refused'    (output: 'ssh: connect to host 10.66.147.191 port 22: Connection refused\n')
22:37:02 DEBUG| Sending command: grep 'MemTotal:' /proc/meminfo
22:37:03 INFO | Assigned 3072M memory to 'avocado-vt-vm1'
22:37:03 INFO | Context: Run sub test 'pause_vm' after hotplug memory device
22:37:05 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 600s)
22:37:09 INFO | Context: Stop the VM
22:37:09 DEBUG| (monitor qmpmonitor1) Sending command 'stop'
22:37:09 DEBUG| Send command: {'execute': 'stop', 'id': 't0t9SPYN'}
22:37:09 INFO | Context: Stop the VM --> Verify the status of VM is 'paused'
22:37:09 INFO | Context: Stop the VM --> Verify the session has no response
22:38:48 INFO | Update MAC(6c:3b:e5:1d:95:79)<->(10.66.144.14)IP pair into address_cache
22:42:14 INFO | Context: Resume the VM
22:42:14 DEBUG| (monitor qmpmonitor1) Sending command 'cont'
22:42:14 DEBUG| Send command: {'execute': 'cont', 'id': 'TODlMvd4'}
22:42:14 INFO | Context: Resume the VM --> Verify the status of VM is 'running'
22:42:14 INFO | Context: Resume the VM --> Re-login the guest
22:42:14 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 600s)
22:42:18 INFO | Context: Run sub test 'pause_vm' after hotplug memory device --> rebooting 'avocado-vt-vm1'
----------------------------------------------------------------------------------
10-smp_1.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.hotplug_memory.during.pause_vm.hotplug.backend_ram.policy_bind.two Finshed 2016-09-28 23:01:05 2016-09-28 23:12:42 PASS 696.800968
debug.log
23:02:19 INFO | Context: (run) --> Hotplug memory 'mem-plug' to VM
23:02:19 DEBUG| (monitor qmpmonitor1) Sending command 'object-add'
23:02:19 DEBUG| Send command: {'execute': 'object-add', 'arguments': {'id': 'mem-plug', 'qom-type': 'memory-backend-ram', 'props': {'policy': 'bind', 'host-nodes': [0], 'size': 1073741824}}, 'id': 'TZyTH96D'}
23:02:19 INFO | Context: (run) --> Hotplug memory 'mem-plug' to VM --> Get hotpluged memory address
23:02:19 DEBUG| (monitor qmpmonitor1) Sending command 'query-memory-devices'
23:02:19 DEBUG| Send command: {'execute': 'query-memory-devices', 'id': 'enAtaYGE'}
23:02:19 DEBUG| (monitor qmpmonitor1) Response to 'query-memory-devices' (re-formated)
23:02:19 DEBUG| (monitor qmpmonitor1)        data:
23:02:19 DEBUG| (monitor qmpmonitor1)            slot: 0
23:02:19 DEBUG| (monitor qmpmonitor1)            node: 1
23:02:19 DEBUG| (monitor qmpmonitor1)            addr: 4294967296
23:02:19 DEBUG| (monitor qmpmonitor1)            memdev: /objects/mem-mem1
23:02:19 DEBUG| (monitor qmpmonitor1)            id: dimm-mem1
23:02:19 DEBUG| (monitor qmpmonitor1)            hotpluggable: True
23:02:19 DEBUG| (monitor qmpmonitor1)            hotplugged: False
23:02:19 DEBUG| (monitor qmpmonitor1)            size: 1073741824
23:02:19 DEBUG| (monitor qmpmonitor1)        type: dimm
23:02:19 INFO | Context: (run) --> Hotplug pc-dimm 'dimm-plug' to VM
23:02:19 DEBUG| (monitor qmpmonitor1) Sending command 'device_add'
23:02:19 DEBUG| Send command: {'execute': 'device_add', 'arguments': OrderedDict([('driver', 'pc-dimm'), ('id', 'dimm-plug'), ('memdev', 'mem-plug')]), 'id': '0XunbCIl'}
23:02:19 INFO | Context: (run) --> Verify memory info
23:02:19 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 240s)
23:02:27 DEBUG| Login timeout expired    (output: '\nLast login: Thu Sep 29 10:53:32 2016 from amd-5400b-4-2.englab.nay.redhat.com\n')
23:02:31 DEBUG| Sending command: grep 'MemTotal:' /proc/meminfo
23:02:31 INFO | Assigned 3072M memory to 'avocado-vt-vm1'
23:02:36 INFO | Context: Stop the VM
23:02:36 DEBUG| (monitor qmpmonitor1) Sending command 'stop'
23:02:36 DEBUG| Send command: {'execute': 'stop', 'id': 'RntnX6ZN'}
23:02:40 INFO | Context: Stop the VM --> Verify the status of VM is 'paused'
23:02:40 INFO | Context: Stop the VM --> Verify the session has no response
23:07:45 INFO | Context: Resume the VM
23:07:45 DEBUG| (monitor qmpmonitor1) Sending command 'cont'
23:07:45 DEBUG| Send command: {'execute': 'cont', 'id': 'GF9N5aF0'}
23:07:45 INFO | Context: Resume the VM --> Verify the status of VM is 'running'
23:07:45 INFO | Context: Resume the VM --> Re-login the guest
23:07:45 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 600s)
23:07:53 INFO | Context: (run) --> rebooting 'avocado-vt-vm1'

**Comment by xutian** [comment]
@xiagao  as I mentioned in comment#0 'hotplug_memory.during.pause_vm.hotplug.backend_file' match case in polarion. other two test case are redundant.

File: qemu/tests/hotplug_mem.py [hunk 1]
```diff
@@ -56,35 +63,50 @@ def get_mem_by_name(self, vm, name):
         devices = filter(None, map(vm.devices.get_by_qid, dev_ids))
         return [_[0] for _ in devices]
 
+    def unplug_memory(self, vm, target_mem):
+        """Unplug the target memory, if the memory not exists,
+           hotplug it, then unplug it
+        """
+        devs = self.get_mem_by_name(vm, target_mem)
+        if not devs and self.params.get("strict") != "yes":
```
Comment (CongLi) [inline qemu/tests/hotplug_mem.py]: Maybe like "strict_check" will be better, only "strict" is not clear.

File: qemu/tests/hotplug_mem.py [hunk 2]
```diff
@@ -56,35 +63,50 @@ def get_mem_by_name(self, vm, name):
...
         target_mem = self.params["target_mem"]
+        stage = self.params.get("stage", "before")
+        sub_test_runner = (
+            stage == 'during' and [
+                self.run_background_test] or [
+                self.run_sub_test])[0]
+        func = getattr(self, "%s_memory" % operation)
+        if not callable(func):
+            raise exceptions.TestError(
+                "Unsupport memory operation '%s'" %
```
Comment (CongLi) [inline qemu/tests/hotplug_mem.py]: s/unsupport/unsupported

**Comment by xutian** [comment]
@CongLi  updated, please help review, thanks.

File: qemu/tests/hotplug_mem.py [hunk 3]
```diff
@@ -56,35 +63,50 @@ def get_mem_by_name(self, vm, name):
...
+        if not callable(func):
+            raise exceptions.TestError(
+                "Unsupported memory operation '%s'" %
+                operation)
+        vm = self.env.get_vm(self.params["main_vm"])
         try:
+            if stage != "after":
+                sub_test = sub_test_runner()
+                func(vm, target_mem)
+                self.check_memory(vm)
```
Comment (suqinhuang) [inline qemu/tests/hotplug_mem.py]: if
else
if
the  syntax is strange

Comment (xutian) [inline qemu/tests/hotplug_mem.py]: we need prepare subtest before hotplug/unplug mem device both before and during, so I have such if else branch. Maybe I need a doc string here.

Thanks your comments.

**Comment by suqinhuang** [comment]
LGTM
Ack

**Comment by xutian** [comment]
merged, thanks @suqinhuang

**Pull request merged**